parla/task_runtime.py

```diff
@@ -1161,6 +1302,106 @@ def _assignment_policy(self, task: Task):
+
+            # First, we calculate data on the device and data to be moved to the device
+            local_data = 0
+            nonlocal_data = 0
+            for parray in task.dataflow.input + task.dataflow.inout:
+                if self._available_resources.parray_is_on_device(parray, device):
+                    local_data += parray.nbytes
+                else:
+                    nonlocal_data += parray.nbytes
+
+            # These values are really big, so I'm normalizing them to the size of the
+            # device memory so my monkey brain can fathom the numbers
+            local_data /= device.resources['memory']
+            nonlocal_data /= device.resources['memory']
+
+            # Figure out whether the task has a dependency running on this device
+            this_device_owns_dependency = False
+            for dependency in task.dependencies:
+                if device in dependency.req.devices:
+                    this_device_owns_dependency = True
+                    break
+
+            # If the best device owns a dependency, but this one doesn't,
+            # AND the best device has more local data, we can skip this device
+            # TODO: Ignore this if the dependency has finished running!
+            if best_device_owns_dependency and not this_device_owns_dependency and best_device_local_data >= local_data:
+                continue
+            
+            # Next we calculate the load-balancing factor
+            # For now this is just a count of tasks on the device queue (TODO: better heuristics later...)
+            dev_load = self._device_compute_task_counts[device]
+
+            # Normalize this too so we have numbers between 0 and 1
+            dev_load /= self._active_task_count
+
+            # TODO: Move these magic numbers somewhere better
+            local_data_weight = 30.0
+            nonlocal_data_weight = 10.0
+            load_weight = 1.0
+
+            # Calculate the suitability
+            suitability = local_data_weight * local_data \
+                        - nonlocal_data_weight * nonlocal_data \
+                        - load_weight * dev_load
+
+            """
+            def myformat(num):
+                return "{:.3f}".format(num)
+
+            print(f"dev={device}", end='   ')
+            print(f"dep={this_device_owns_dependency}", end='   ')
+            print(f"local={myformat(local_data)}", end='   ')
+            print(f"nonlocal={myformat(nonlocal_data)}", end='   ')
+            print(f"load={myformat(dev_load)}", end='   ')
+            print(f"suit={myformat(suitability)}")
+            """
+
+            # Update whether or not this is the most suitable device
+            if max_suitability is None or suitability > max_suitability:
+                max_suitability = suitability
+                best_device = device
+                best_device_owns_dependency = this_device_owns_dependency
```

Comment:
Could you plesae describe 'own' meaning in this context?

Reply:
The device "owns" the dependency task, or in other words, the dependency task is mapped to the device.